ENH: adds np.nancumsum and np.nancumprod

pwolfram opened the pull request:
This PR adds an implementation of `nancumsum` and `nancumprod`.

The actual function is a two-liner adapted from `nansum` and `nanprod`.

Its structure is adapted from PR: #5418 ( a minor typo in the doc string from this PR is fixed too)

pwolfram commented:
cc @shoyer and https://github.com/pydata/xarray/issues/791

shoyer commented:
Just a note -- since this is new API, somebody might ask you to write the numpy-discussions mailing list. But I think this should be pretty uncontroversial :+1:.

shoyer commented:
Consistency checks with integer arrays are great, but these also need tests on arrays with NaNs.

pwolfram commented:
Thanks @shoyer, I accidentally left that out.  I updated this to include tests on arrays with NaNs following `nanprod`'s lead.  Does this work?

charris commented:
Could you put the documentation commit in this PR as well? Note that you can keep adding commits to the PR, and later clean up the history with `git rebase -i` if needed.

pwolfram committed:
ENH: adds np.nancumsum and np.nancumprod

This PR adds an implementation of `nancumsum` and `nancumprod`.

The actual function is a two-liner adapted from `nansum`.

Its structure is adapted from PR: https://github.com/numpy/numpy/pull/5418/

pwolfram commented:
@charris, done.  @shoyer, I've also added the testing capability as you requested.  Thanks!

shoyer commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +609,145 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+
+    One is returned for slices that are all-NaN or empty.
+
+    .. versionadded:: 1.11.0
```
We're already at the release candidate stage for 1.11, so this will make it into 1.12

pwolfram commented:
@charris and @shoyer, is there anything else that needs done on this PR?

seberg commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +609,145 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    Return the cumulative sum of array elements over a given axis treating Not a
+    Numbers (NaNs) as zero.
+
+    One is returned for slices that are all-NaN or empty.
```
I guess this is supposed to say zero?

pwolfram replied:
@seberg, thanks for finding that typo (should have been plural form of verb too, not singular-- same also for nancumprod)

njsmith commented:
Mailing list thread: https://mail.scipy.org/pipermail/numpy-discussion/2016-March/075169.html

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -548,7 +551,7 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     a : array_like
-        Array containing numbers whose sum is desired. If `a` is not an
+        Array containing numbers whose product is desired. If `a` is not an
         array, a conversion is attempted.
```
While you're at it, nans are treated like "one", not "zero" on line 542/545.

madphysicist replied:
Also, if I am not mistaken, `axis`, two lines below, accepts a tuple now. The docs should be updated according to `np.prod`.

jaimefrio replied:
Not for the `.accumulate()` method used by the `cumxxx` functions: only one axis.

madphysicist replied:
My mistake. By the way, would it make sense to apply the `_ureduce` function from `numpy.lib.function_base` to other places, like most of `numpy.core.from_numeric`? That way we could add multi-axis support without waiting for the `gufunc` redesign that seems to be coming.

jaimefrio replied:
Which functions do you have in mind? I don't see any obvious place where `_ureduce` could help...

madphysicist replied:
`ptp`, `argmin`, `argmax`. Also, the following might benefit as well, but the internal order of the array would matter: `partition`, `argpartition`, `sort`, `argsort`, `searchsorted`, `cumsum`, `cumprod`. I was even thinking that it might be worth allowing `ravel` to do a partial ravel along a subset of the axes (basically like `_ureduce` does), but that is probably going to require some major rewriting to do properly.

pwolfram replied:
Thanks @madphysicist, change made by `8fd1ee8` fixes the typo.  However, for your second comment-- I am not able to get the behavior you suggest for `np.prod` with `axis`, e.g.,

```
In [1]: import numpy as np

In [2]: a = np.random.randn(10,10)

In [3]: np.prod(a,axis=(1,2))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-56314d82b79d> in <module>()
----> 1 np.prod(a,axis=(1,2))

/Users/pwolfram/anaconda/lib/python2.7/site-packages/numpy/core/fromnumeric.pyc in prod(a, axis, dtype, out, keepdims)
   2544 
   2545     return _methods._prod(a, axis=axis, dtype=dtype,
-> 2546                           out=out, **kwargs)
   2547 
   2548 

/Users/pwolfram/anaconda/lib/python2.7/site-packages/numpy/core/_methods.pyc in _prod(a, axis, dtype, out, keepdims)
     33 
     34 def _prod(a, axis=None, dtype=None, out=None, keepdims=False):
---> 35     return umr_prod(a, axis, dtype, out, keepdims)
     36 
     37 def _any(a, axis=None, dtype=None, out=None, keepdims=False):

ValueError: 'axis' entry is out of bounds
```

Am I missing something?

madphysicist replied:
Only that I made a mistake. Axis only accepts one value here. You were absolutely right.

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +609,145 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    Return the cumulative sum of array elements over a given axis treating Not a
+    Numbers (NaNs) as zero.
+
+    Zeros are returned for slices that are all-NaN or empty.
```
Mention that the sum does not change when nans are encountered and that any leading nans are replaced by zeros.

pwolfram replied:
@madphysicist, thanks, should be fixed by `41adc0f`

madphysicist replied:
Have you pushed that commit yet?

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +609,145 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+def nancumprod(a, axis=None, dtype=None, out=None):
+    """
+    Return the cumulative product of array elements over a given axis treating Not a
+    Numbers (NaNs) as zero.
```
As one. Hold the line!

pwolfram replied:
Thanks @madphysicist, see `1854184`

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +609,145 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    -------
+    nancumprod : ndarray
+        A new array holding the result is returned unless `out` is
+        specified, in which case a reference to out is returned.
```
Either "it is returned" or backticks around the second "out".

madphysicist replied:
By the way, not your scope, but it might be worth adding the bit about out being returned to nansum and nanprod.

pwolfram replied:
Thanks @madphysicist, `7931758` should make this change choosing simpler "it is returned" for `nancumsum` and `nancumprod`.

pwolfram replied:
@madphysicist, added return to `nansum` and `nanprod` in `84acc4c` to improve doc.

madphysicist commented:
Aside from the comments about docs, LGTM. And by all means feel free to ignore the comments about functions other than your own.

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -685,7 +685,7 @@ def nancumprod(a, axis=None, dtype=None, out=None):
     Numbers (NaNs) as zero.
 
-    One is returned for slices that are all-NaN or empty.
+    Ones are returned for slices that are all-NaN or empty.
```
Thanks. One more two lines up: "as zero" -> "as one".

pwolfram commented:
@madphysicist, I've pushed commits to address your comments (including for `nansum` and `nanprod`).  Please let me know if you see any further room for improvement.  Thanks!

pwolfram commented:
@madphysicist, actually there are more following a page refresh... hold on a bit.  I'll let you know when the others are pushed.

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -493,7 +496,11 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    nansum_along_axis : ndarray.
+        A new array holding the result is returned unless `out` is
+        specified, in which it is returned. The
+        result has the same size as `a`, and the same shape as `a` if
```
Sorry to nitpick, but the line lengths look funny here.

pwolfram replied:
Agreed I wondered about that myself but didn't want to depart from convention.  Fixed in `29cca2d`

pwolfram replied:
Also in `6ae90ec`

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +615,147 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+
+    Returns
+    -------
+    nancumsum_along_axis : ndarray.
```
See comment above about `along_axis`. `nancumsum` is fine without the suffix. Also, it may not be along an axis for most cases (axis is None by default).

pwolfram replied:
Fixed in `56aefae`

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +615,147 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    -------
+    nancumsum_along_axis : ndarray.
+        A new array holding the result is returned unless `out` is
+        specified, in which it is returned. The
```
Line lengths.

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +615,147 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    -----
+    Numpy integer arithmetic is modular. If the size of a product exceeds
+    the size of an integer accumulator, its value will wrap around and the
+    result will be incorrect. Specifying ``dtype=double`` can alleviate
```
Either `np.double` or "an inexact type" would probably be better.

madphysicist commented on numpy/lib/nanfunctions.py:
```diff
@@ -606,6 +615,147 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+    -----
+    Numpy integer arithmetic is modular. If the size of a product exceeds
+    the size of an integer accumulator, its value will wrap around and the
+    result will be incorrect. Specifying ``dtype=double`` can alleviate
```
Either `np.double` or "an inexact type" would probably be better. (Same comment below)

pwolfram replied:
Made replacement to `np.double` in `a6ad716` throughout file.

madphysicist commented:
By the way, if you are doing multiple commits with documentation changes, you can add "[ci skip]" (without the quotes) anywhere in the commit message to prevent Travis (and I think AppVeyor) from rerunning all the tests.

pwolfram commented:
Thanks @madphysicist for the tip about adding `[ci skip]` to the commit.  Do you want me to retroactively apply this?  I'm assuming that it is probably too late to save time for the CI.

madphysicist commented:
I don't think you can do much to stop the jobs once they have been scheduled. I wouldn't worry about it too much though.

madphysicist commented:
Looks like you've fixed up all of my nitpicks, so LGTM. Very nice work.

pwolfram commented:
Thanks @madphysicist for the big help!

madphysicist commented:
I always enjoy getting thanked when the other guy is doing all the work :)

shoyer commented on numpy/lib/nanfunctions.py:
```diff
@@ -508,7 +515,7 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
 
     Numpy integer arithmetic is modular. If the size of a sum exceeds the
```
This entire note seems out of place for nan functions. There isn't an integer NaN, so even though the functions work on integer arrays, the warning is misleading. If anything, these functions should advice using the normal `np.sum` function for integers, rather than warning about integer overflow.

shoyer commented on numpy/lib/nanfunctions.py:
```diff
@@ -508,7 +515,7 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     Numpy integer arithmetic is modular. If the size of a sum exceeds the
     size of an integer accumulator, its value will wrap around and the
-    result will be incorrect. Specifying ``dtype=double`` can alleviate
+    result will be incorrect. Specifying ``dtype=np.double`` can alleviate
```
If you're going to keep this language, don't keep the word "double" which is non-standard from a numpy perspective. "float" or "float64" would be the more standard terms.

pwolfram replied:
@shoyer thank you for taking another look and finding this.  Perhaps the principle of minimalism should be employed and 

```
    Numpy integer arithmetic is modular. If the size of a sum exceeds the
    size of an integer accumulator, its value will wrap around and the
    result will be incorrect. Specifying ``dtype=np.double`` can alleviate
    that problem.
```

be removed?  It seems like a bit of a corner case that, as you mention, probably would be covered by the `np.sum` documentation anyway. 

@charris, can you please comment on this because you originally wrote this comment and I ended up just doing a copy-paste style operation.  This language exists for `np.nansum`, `np.nanprod`, `np.nancumsum`, and `np.nancumprod` and should be edited anyway because you'll note that it isn't quite pertinent to each function (e.g., use of product for `np.nancumsum`.)  I can edit it once we reach consensus on the right decision here.

pwolfram replied:
After thinking about this more I think we should just remove the text because it is a secondary issue anyway, is covered by the `np.sum` documentation, and is bloating the documentation.  This is done in `3509317`.

pwolfram commented:
Thank you everyone.  If there are no more changes I can squash the commits so this can be merged.

madphysicist commented:
LGTM

pwolfram commented:
Commits squashed-- this is ready to merge

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,101 @@ def test_empty(self):
+    def test_allnans(self):
+        # Check for FutureWarning
+        for f, tgt_value in zip(self.nanfuncs, [0, 1]):
+            with warnings.catch_warnings(record=True) as w:
```
It's cleaner to use [`assert_warns`](https://github.com/numpy/numpy/blob/a6efa09837b3a71eb6f6450b7dcf08c7381e75d0/numpy/testing/utils.py#L1683) and `assert_no_warnings` instead, e.g.,

```
with assert_warns(FutureWarning):
    # ....
```

seberg replied:
what is assert no warnings anyway? I personally want warnings to be errors always inside the test (unless it is release version) to be honest. Asserting no warning seems like a broken concept. If there are warnings, you have to catch them, if there are not, you don't. If you have to mix it, you are screwed, unless you merge my new PR, heh....

seberg replied:
well, not screwed, it just can get quite annoying.

shoyer replied:
In general, I agree. I vaguely recall discussing this at one point -- is there a good reason why warnings aren't automatically errors on github?

That said, I think there is also something to be said for explicitly indicating there shouldn't be warnings in code that might otherwise issue them.

seberg replied:
Yeah, sure, nothing wrong with this here. Just thought it was a bit
funny and was wondering if someone disagrees ;). I probably should make
these context managers use my context manager instead of catch
warnings, if we merge it. Though I think I would do that after the
other is in.

shoyer replied:
*your* context manager? but *I* wrote `assert_no_warnings` -- which I think is the right one here :)

seberg replied:
@shoyer see gh-7099, the assertion context managers are probably nice, especially as context manager. But they use `catch_warnings` inside, and `catch_warnings` is not as helpful as it could be. In gh-7099 I propose a pimped up `catch_warnings` context which actually makes things easier (nested catch warnings, silence warning without catching, still printing warning in release mode). The assertion context managers could use that one internally to be more specific.

Overall, I want our tests to be able to reliably test all warnings in all python versions, and creating a new fancy context manager, while annoying, seemed like the easiest approach.

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,101 @@ def test_empty(self):
+        for f in self.nanfuncs:
+            d = np.ones((3, 5, 7, 11))
+            # Randomly set some elements to NaN:
+            w = np.random.random((4, 200)) * np.array(d.shape)[:, None]
```
I prefer:
```python
rs = np.random.RandomState(0)  # randomized tests should always use a seed
d[rs.rand(d.shape) < 0.5] = np.nan  # more obvious then your broadcasting tricks
```

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -22,6 +22,18 @@
          np.array([0.1610, 0.1859, 0.3146])]
 
+# Rows of _ndat with nans converted to ones
+_rdat_ones = [np.array([0.6244 , 1.0     , 0.2692 , 0.0116 , 1.0    , 0.1170])  ,
```
This is technically a pep8 spacing violation, so it's probably worth correcting.

I know this is a little pedantic, but lots of developers setup their editors to run lint automatically, which I recommend also adopting. In pandas, violating PEP8 now causes a test suite error but in NumPy we still rely on manual checks.

shoyer commented on doc/release/1.12.0-notes.rst:
```diff
@@ -6,6 +6,9 @@ This release supports Python 2.7 and 3.4 - 3.5.
 
+Details of these improvements can be found below.
+
+* Addition of `nancumsum` and `nancumprod` to the set of nanfunctions.
```
Better to add these to "new features" with a brief description -- we'll fill out highlights shortly before cutting the release.

shoyer commented:
Sorry, I only now took a careful look at your unit tests. I think the coverage is pretty good now, though I do have some suggestions for cleaning up the tests a little bit.

If you have a principled reason for making three separate commits here, that's OK I guess, but right now it seems a little strange -- especially how one of the commits documenting nancumsum/nanprodsum goes in before the commit adding them! Personally I would squash everything down to one commit.

pwolfram commented:
@shoyer, the old style of warning is employed throughout in other parts of the `test_nanfunctions.py`, e.g., with `warnings.catch_warnings(record=True) as w:`.  Is it ok to leave this as is?  Presumably this should be cleaned up but this is probably a separate PR.

shoyer commented:
@pwolfram I think it's fine to not update the style for old tests (refactoring old tests is always a little dangerous) but the clarity of new tests would probably benefit from using the context managers.

pwolfram commented:
@shoyer, agreed. Did my force pushed changes meet this goal appropriately or do you see room for further improvement?  Thanks!

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,94 @@ def test_empty(self):
+    def test_allnans(self):
+        for f, tgt_value in zip(self.nanfuncs, [0, 1]):
+            with assert_no_warnings():
+                res = f([np.nan]*3, axis=None)
```
per pep8, you're missing spaces around `*` and `%` in these tests.

pwolfram replied:
`pylint` didn't seem to find these and the closest reference I could find was https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements.  Is there a better one?

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,94 @@ def test_empty(self):
+            with assert_no_warnings():
+                res = f(d, axis=None)
+                assert_equal(res.shape, (1155,))
+                res = f(d, axis=0)
```
use a loop -- `for axis in range(4):`

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,94 @@ def test_empty(self):
+            # Randomly set some elements to NaN:
+            rs = np.random.RandomState(0)
+            d[rs.rand(*d.shape) < 0.5] = np.nan
+            with assert_no_warnings():
```
Per @seberg's suggestion, maybe we don't need the `assert_no_warnings()` check arounds block. I still think it serves a purpose above when testing on the all-NaN values.

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,94 @@ def test_empty(self):
+
+    def test_allnans(self):
+        for f, tgt_value in zip(self.nanfuncs, [0, 1]):
+            with assert_no_warnings():
```
add a comment that unlike other nan-functions, sum/prod/cumsum/cumprod don't warn on all nan input.

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,94 @@ def test_empty(self):
+    def test_empty(self):
+        for f, tgt_value in zip(self.nanfuncs, [0, 1]):
+            mat = np.zeros((0, 3))
+            tgt = tgt_value*np.ones((0, 3))
```
more pep8... I do recommend installing pylint :)

pwolfram replied:
Perfect, went with the vim's https://github.com/scrooloose/syntastic plugin and `pylint`.  :+1:

madphysicist replied:
I noticed that scipy and matplotlib (and probably many others) have a pep8 Travis test. Just out of curiosity, why doesn't numpy?

pwolfram commented:
@shoyer, quick question-- `pylint` didn't necessarily highlight the issues with `*` or `%`, either in vim or via the console run with `pylint`.  Is there something I'm missing here, e.g., do you do something different?  The recommendation is in the spec at https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements so it is surprising `pylint` doesn't catch it.

pwolfram commented:
Changes made, thanks for the keen attention to detail and for helping me improve the code @shoyer and @seberg!

shoyer commented:
Reading PEP8 more carefully, it looks like I'm wrong about spacing around binary operators:
https://www.python.org/dev/peps/pep-0008/#other-recommendations

Hmm....

pwolfram commented:
@shoyer, reverted spacing back on `*` because I don't think this makes the code easier to read but left it for `%` because it adds clarity in accordance with the pep8 guidelines.

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -22,6 +22,18 @@
          np.array([0.1610, 0.1859, 0.3146])]
 
+# Rows of _ndat with nans converted to ones
+_rdat_ones = [np.array([0.6244, 1.0, 0.2692, 0.0116, 1.0, 0.1170]),
```
I think this should probably be all one big array -- it's only not one array for `_rdat` because `_rdat` is ragged

shoyer commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,87 @@ def test_empty(self):
+
+    def test_result_values(self):
+        tgt = [np.cumprod(d) for d in _rdat_ones]
+        res = np.nancumprod(_ndat, axis=1)
```
Once you make `_rdat_ones` and `_rdat_zeros` arrays, you could add a test looping over all axes `(-2, -1, 0, 1, None)`. That would be a nice sanity check -- especially because you currently have no tests for negative axes.

shoyer commented:
@pwolfram  I think this is very close, though a test for negative axes would be nice. Otherwise this looks good to me. Sorry for leading your astray on pep8!

pwolfram commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,88 @@ def test_empty(self):
+            assert_almost_equal(res, tgt)
+            tgt = np.cumsum(_ndat_zeros,axis=axis)
+            res = np.nancumsum(_ndat, axis=axis)
+            assert_almost_equal(res, tgt)
```
@shoyer, what this what you were thinking?

shoyer replied:
yes, exactly

pwolfram commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -388,6 +410,89 @@ def test_empty(self):
+                res = nf(mat, axis=axis, out=resout)
+                assert_almost_equal(res, resout)
+                assert_almost_equal(res, tgt)
+
```
@shoyer, I also generalized the test here for consistency and greater coverage.

pwolfram commented:
@shoyer, please see above tests added over the negative axis.  Thanks for helping refine this!

pwolfram commented:
Also, btw @shoyer, no worries about the pep8 confusion.  I installed the plugin for vim and I have a clearer idea of formatting standards so it was certainly value-added!

shoyer commented:
OK, this looks good to me. I'll merge this once tests pass unless anyone speaks up with objections...

pwolfram commented:
Thanks @shoyer!
